<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Promise</title>
</head>
<body>
    <script>
        //resolve, reject, all, race
        //resolve == then, reject == catch

        // let promise = new Promise((resolve, reject) => {
        //     resolve(1111);
        // });
        // promise.then(response => {
        //     console.log(response);
        //     console.log("成功")
        // }).catch(error => {
        //     console.log(error);
        //     console.log("失败")
        // });

        //链式
        // function promise1(status) {
        //     return new Promise((reslove, reject)=>{
        //         if(status) {
        //             console.log("第1个promise成功");
        //             reslove("第1个promise返回数据成功");
        //         }
        //         if(!status) {
        //             console.log("第2个promise失败");
        //             reslove("第2个promise返回数据失败");
        //         }
        //     })
        // }
        // promise1(true).then(response => {
        //     console.log(response);
        // }).catch(error => {
        //     console.log(error);
        // })

        function promise1(status) {
            return new Promise((reslove, reject)=>{
                if(status) {
                    console.log("第1个promise成功");
                    reslove("第1个promise返回数据成功");
                }
                if(!status) {
                    console.log("第1个promise失败");
                    reject("第1个promise返回数据失败");
                }
            })
        }
        function promise2(status) {
            return new Promise((reslove, reject)=>{
                if(status) {
                    console.log("第2个promise成功");
                    reslove("第2个promise返回数据成功");
                }
                if(!status) {
                    console.log("第2个promise失败");
                    reject("第2个promise返回数据失败");
                }
            })
        }
        // promise1(true).then(response => {
        //     console.log(response);
        //     return promise2(false);
        // }).then((response) => {
        //     console.log(response);
        // }).catch(error => {
        //     console.log(error);
        // });

        //all里面数组全部成功，才会执行then回调
        // Promise.all([promise1(true), promise2(false)]).then((response)=>{
        //     console.log("全部调用成功");
        // }).catch((error)=>{
        //     console.log("有一些失败");
        // });

        //race数组,只要有一个返回resolve,就代表成功，就会回调then.但是,如果第一个返回reject,那么就失败了。无论失败成功，都会往下执行
        Promise.race([promise1(false), promise2(true)]).then((response)=>{
            console.log("成功");
        }).catch((error)=>{
            console.log("失败了");
        });
    </script>
</body>
</html>